### PR TITLE
EOS-20347 PAM AUTH POC

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Backend

## Problem Statement
<pre>
Story Ref (if any):
</pre>
[EOS-20347](https://jts.seagate.com/browse/EOS-20347)
## Unit testing on RPM done
<pre>
No
</pre>
## Problem Description
<pre>
PAM Authentication POC
</pre>
## Solution
<pre>
Created pam_cortx module which will authenticate user through csm rest api.
Updated readme file with steps to integrate pam_cortx module on a system. 
</pre>
## Unit Test Cases
<pre>
1. Logging In with CSM user successful after doing ssh to a pam_cortx configured VM(3385)
931949 is CMS user with manage role as well as it is a  LDAP user.
</pre>
image
<pre>
Logs(/var/log/secure) on 3385 VM:
</pre>
logging In with CSM User

<pre>
2. Logging In with CSM user Failed after doing ssh to a pam_cortx configured VM.
 After CSM Authentication failure, for the same user authentication has been performed against next rules in the stack 
 1. Linux User 2. Ldap User
</pre>
logging In with CSM User failed

Signed-off-by: rohitkolapkar <rohit.j.kolapkar@seagate.com>

### CODE REVIEW

**Comment by rohitkolapkar** [comment]
If user ssh on a vm which is configured with pam_cortx module then authentication sequence is,
1. CSM authentication using REST API
2. Linux authentication
3. LDAP authentication


image

This sequence can be altered by editing **/etc/pam.d/password-auth**
[sample_configuration_password-auth.txt](https://github.com/Seagate/cortx-manager/files/6492945/sample_configuration_password-auth.txt)

**Comment by rohitkolapkar** [comment]
As cortx users/s3 users are going to be LDAP users, we can configure  **/etc/nslcd.conf**  file to map the user with **cortxcli** **loginShell** after user login successfully. 

[sample_nslcd-conf.txt](https://github.com/Seagate/cortx-manager/files/6493059/sample_nslcd-conf.txt)

File: experiments/pam/pam_cortx.c [hunk 1]
```diff
@@ -72,11 +72,12 @@ static int login(pam_handle_t *pamh, const char *pUsername, const char *pPasswor
 	json_object_object_add(jobj, "password", json_object_new_string(pPassword));
 	/* set curl options */
 	curl_easy_setopt(pCurl, CURLOPT_URL, pUrl);
+	curl_easy_setopt(pCurl, CURLOPT_FAILONERROR, TRUE);
```
Comment (tb-s) [inline experiments/pam/pam_cortx.c:75]: We are not checking for errors in any of the `curl_easy_setopt()` calls (to say the least).  Any function that might return an error code must be checked.

File: experiments/pam/pam_cortx.c [hunk 2]
```diff
@@ -49,14 +49,14 @@ static int login(pam_handle_t *pamh, const char *pUsername, const char *pPasswor
 {
     //Login Function for CSM Initialization.
 	pam_syslog(pamh, LOG_INFO, "Verify CSM User %s \n", pUsername);
-	const char *pUrl = "http://localhost:8101/api/v1/login?debug";
+	const char *pUrl = "http://localhost:28101/api/v1/login";
 	CURL *pCurl = curl_easy_init();
 	struct curl_slist *headers = NULL;
 	int res = -1;
 	static const char *header_file_name = "/var/log/response_headers.out";
```
Comment (tb-s) [inline experiments/pam/pam_cortx.c:56]: - `/var/log` does not seem a good location to these files; why can't we use `/tmp` instead?
- A randomized filename would help mitigate (not to fix, I believe) the issue pointed out by Codacy.  See `mkstemp(3)`.

File: experiments/pam/pam_cortx.c [hunk 3]
```diff
@@ -85,7 +86,7 @@ static int login(pam_handle_t *pamh, const char *pUsername, const char *pPasswor
 	}
 
 	/* open the body file */
-	body_file = fopen(body_file_name, "wb");
+	body_file = fopen(body_file_name, "w");
```
Comment (tb-s) [inline experiments/pam/pam_cortx.c:89]: Will we reuse the file somewhere else?  If not, we should open it with `"w+"` and delete the file before the function returns.  Same for the other `fopen()` call.

File: experiments/pam/pam_cortx.c [hunk 4]
```diff
@@ -49,14 +49,14 @@ static int login(pam_handle_t *pamh, const char *pUsername, const char *pPasswor
 {
     //Login Function for CSM Initialization.
 	pam_syslog(pamh, LOG_INFO, "Verify CSM User %s \n", pUsername);
-	const char *pUrl = "http://localhost:8101/api/v1/login?debug";
+	const char *pUrl = "http://localhost:28101/api/v1/login";
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply experiments/pam/pam_cortx.c]: I made it as a static.  @tb-s

File: experiments/pam/pam_cortx.c (hunk 3 above)
Comment (rohitkolapkar) [reply experiments/pam/pam_cortx.c]: @tb-s  I am not sure, are we going to reuse the file or not. Task was to store response headers and response body in some file.

File: experiments/pam/pam_cortx.c (hunk 1 above)
Comment (rohitkolapkar) [reply experiments/pam/pam_cortx.c:75]: @tb-s Thanks for pointing out, as a part of this POC, we were trying to explore how to use REST api in PAM authentication. This is just a sample code for that. we are going to rewrite this entire PAM module in upcoming sprint.

File: experiments/pam/pam_cortx.c (hunk 3 above)
Comment (tb-s) [reply experiments/pam/pam_cortx.c]: If we are truncating the file when we open it, chances are we are not reusing it unless this is part of a sequence of commands with a shared dependency on that file.

**Pull request merged**